code/cfile/cfilecompression.cpp

```diff
@@ -0,0 +1,219 @@
+		Assertion(cf->size > 4, "Invalid filesize, compressed file is possibly in the wrong format.");
+		Assertion(cf->compression_info.block_size > 16, "Invalid block size, compressed file is possibly in the wrong format.");
+	}
+	#endif
+	cf->compression_info.decoderBuffer = (char*)malloc(LZ4_compressBound(cf->compression_info.block_size));
+	cf->compression_info.lastDecBlockPos = 0;
+	cf->compression_info.lastDecBytes = 0;
+	lz41_load_offsets(cf);
+}
+
+void lz41_load_offsets(CFILE* cf)
+{
+	int maxBlocks = (int)cf->size + 8 / cf->compression_info.block_size;
+	cf->compression_info.offsets = (int*)malloc(maxBlocks);
+	int block;
+	int* offsetsPtr = cf->compression_info.offsets;
+
+	fso_fseek(cf, (-4 * (cf->compression_info.numOffsets + 1)) - (sizeof(int)*2), SEEK_END);
+	for (block = 0; block <= cf->compression_info.numOffsets; ++block)
+	{
+		auto bytes_read = fread(offsetsPtr++, sizeof(int), 1, cf->fp);
+		Assertion(bytes_read == 1, "Error reading offset list.");
+	}
+}
+
+size_t lz41_stream_random_access(CFILE* cf, char* bytes_out, size_t offset, size_t length)
+{
+	LZ4_streamDecode_t lz4StreamDecode_body;
+	LZ4_streamDecode_t* lz4StreamDecode = &lz4StreamDecode_body;
+	/* The blocks (currentBlock to endBlock) contain the data we want */
+	size_t currentBlock = offset / cf->compression_info.block_size;
+	size_t endBlock = ((offset + length - 1) / cf->compression_info.block_size) + 1;
+	size_t written_bytes = 0;
+
+	if (cf->compression_info.numOffsets <= (int)endBlock)
+		return (size_t)LZ41_OFFSETS_MISMATCH;
+
+	/* Seek to the first block to read, if it matches the cached block, search the next one instead */
+	if (cf->compression_info.lastDecBlockPos == cf->compression_info.offsets[currentBlock] && currentBlock + 1 <= endBlock)
+		fso_fseek(cf, cf->compression_info.offsets[currentBlock+1], SEEK_SET);
+	else
+		fso_fseek(cf, cf->compression_info.offsets[currentBlock], SEEK_SET);
+	
+	offset = offset % cf->compression_info.block_size;
+
+	/* Start decoding */
+	for (; currentBlock < endBlock; ++currentBlock)
+	{
+		if (cf->compression_info.lastDecBlockPos != cf->compression_info.offsets[currentBlock])
+		{
+			char* cmpBuf = (char*)malloc(LZ4_compressBound(cf->compression_info.block_size));
+			/* The difference in offsets is the size of the block */
+			int cmpBytes = cf->compression_info.offsets[currentBlock + 1] - cf->compression_info.offsets[currentBlock];
+			auto bytes_read = fread(cmpBuf, cmpBytes, 1, cf->fp);
+			Assertion(bytes_read == 1, "Error reading from compressed file.");
+
+			const int decBytes = LZ4_decompress_safe_continue(lz4StreamDecode, cmpBuf, cf->compression_info.decoderBuffer, cmpBytes, cf->compression_info.block_size);
+			if (decBytes <= 0)
+				return (size_t)LZ41_DECOMPRESSION_ERROR;
+
+			/* Write out the part of the data we care about */
+			size_t blockLength = ((length) < ((decBytes - offset)) ? (length) : ((decBytes - offset)));
```

Comment:
You can strip a few parenthesis here to make the expression easier to read:
`(length < (decBytes - offset) ? length : (decBytes - offset));`